torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -2172,29 +2172,62 @@ void PropagateInputShapes(const std::shared_ptr<Graph>& graph) {
 
 namespace {
 
-void EraseShapeInformation(at::ArrayRef<Value*> vals) {
+using TypeCache = std::unordered_map<TypePtr, TypePtr>;
+
+TypePtr getOrCreateUnshapedType(TypePtr type, TypeCache& cache);
+
+TypePtr unshapedTypeImpl(TypePtr type, TypeCache& cache) {
+  if (type->isSubtypeOf(TensorType::get())) {
+    return TensorType::get();
+  }
+  std::vector<TypePtr> unshaped_contained_types;
+  for (const auto& type: type->containedTypes()) {
+    unshaped_contained_types.push_back(getOrCreateUnshapedType(type));
+  }
+  return type->withContained(unshaped_contained_types);
+}
+
+TypePtr getOrCreateUnshapedType(TypePtr type, TypeCache& cache) {
+  auto maybe_cached_type = cache.find(type);
+  if (maybe_cached_type != cache.end()) {
+    return maybe_cached_type->second;
+  }
+  auto unshaped_type = unshapedTypeImpl(type);
+  cache[type] = unshaped_type;
```

Comment:
nitpick: you could `insert` here, slightly more efficient and more canonical or the first-time insertions.